AngularSelector (for Angular 2+)

miherlosev opened the pull request:
@VasilyStrelyaev first
then @kirovboris @AlexanderMoskovkin 

Don't merge until testcafe 0.18.0 is released.
There are errors in hammerhead that fixed only in testcafe@dev (v0.18.0-dev20170921)

Commit:
initial implementation

Commit:
small fixes

Commit:
smal remark

VasilyStrelyaev commented on README.md:
```diff
@@ -10,88 +11,31 @@ npm install testcafe-angular-selectors
 ## Usage
 
+The module includes separate helpers for AngularJS(1.x) and Angular(2+) applications.
+For AngularJS applications need to use `AngularJSSelector` that contains a set of static methods to search by specified binding (`byModel`, `byBinding` and etc.).
```
```
This module includes separate helpers for AngularJS(1.x) and Angular(2+) applications.

For AngularJS applications, you need to use `AngularJSSelector` that contains a set of static methods to search by the specified bindings (`byModel`, `byBinding` and etc.).
```

VasilyStrelyaev commented on README.md:
```diff
@@ -10,88 +11,31 @@ npm install testcafe-angular-selectors
 
+The module includes separate helpers for AngularJS(1.x) and Angular(2+) applications.
+For AngularJS applications need to use `AngularJSSelector` that contains a set of static methods to search by specified binding (`byModel`, `byBinding` and etc.).
+Example
```
```

**Example**

```

(make it a separate paragraph)

VasilyStrelyaev commented on README.md:
```diff
@@ -10,88 +11,31 @@ npm install testcafe-angular-selectors
 
-### byModel
-Find elements by 'ng-model' expression
+For Angular application the module provides the possibility to select html element by Angular's component selector or nested component selectors.
```
```
For Angular applications, this module provides the capability to select an HTML element by an Angular's component selector or nested component selectors.
```

VasilyStrelyaev commented on README.md:
```diff
@@ -10,88 +11,31 @@ npm install testcafe-angular-selectors
-Find elements by 'ng-model' expression
+For Angular application the module provides the possibility to select html element by Angular's component selector or nested component selectors.
+You can find more information about Angular's component selector in [angular.io documentation topic](https://angular.io/api/core/Component).
+Also module provides the `waitForAngular` helper method. It will wait until Angular's component tree is loaded. 
```
```

You can find more information about Angular's component selector in the [angular.io documentation topic](https://angular.io/api/core/Component).

Also, this module provides the `waitForAngular` helper method. Use it to wait until Angular's component tree is loaded. 
```

VasilyStrelyaev commented on angular-selector.md:
```diff
@@ -0,0 +1,76 @@
+Ensure that your application bootstrapped in development mode.
+By default, this is true, unless code contains [enableProdMode](https://angular.io/api/core/enableProdMode) method call.
+
+If your application run in production mode then `AngularSelector` will not work.
```
```
## Prerequisites  
Ensure that your application is bootstrapped in the development mode.
By default, this is true, unless the code contains the [enableProdMode](https://angular.io/api/core/enableProdMode) method call.

If your application runs in the production mode, you won't be able to use `AngularSelector`.
```

VasilyStrelyaev commented on angular-selector.md:
```diff
@@ -0,0 +1,76 @@
+## Usage
+
+### WaitForAngular
+To wait until Angular'component tree was loaded need to add `waitForAngular` method into fixture's `beforeEach` hook.
```
```
To wait until the Angular's component tree is loaded, add the `waitForAngular` method into fixture's `beforeEach` hook.
```

VasilyStrelyaev commented on angular-selector.md:
````diff
@@ -0,0 +1,76 @@
+```
+Sometime Angular's component tree is very large or your application has a large startup time.
+In this case the assertion `await t.expect(waitForAngular()).ok();` may failed.
+To prevent this you can increase timeout for assertion.
````
```
Sometimes the Angular's component tree is very large or your application takes a lot of time to start up.
In this case, the assertion `await t.expect(waitForAngular()).ok();` may fail.
To prevent this, you can increase timeout for the assertion.
```

VasilyStrelyaev commented on angular-selector.md:
````diff
@@ -0,0 +1,76 @@
+```
+
+### AngularSelector
+`AngularSelector` allows to select html element by Angular's component selector or nested component selectors.
````
```
`AngularSelector` allows you to select an HTML element by Angular's component selector or nested component selectors.
```

VasilyStrelyaev commented on angular-selector.md:
````diff
@@ -0,0 +1,76 @@
+```
+The rootAngular variable will contain the `<my-app>` element.
+
+> If your application has multiple root then `AngularSelector` returns first one that will be returned by `window.getAllAngularRootElements` function
````
```
> If your application has multiple roots, `AngularSelector` will return the first root returned by the `window.getAllAngularRootElements` function.
```

VasilyStrelyaev commented on angular-selector.md:
````diff
@@ -0,0 +1,76 @@
+
+```
+
+see more examples [here](test/angular-selector-test.js);
````
```
See more examples [here](test/angular-selector-test.js).
```

VasilyStrelyaev commented on angularJS-selector.md:
```diff
@@ -0,0 +1,85 @@
+# AngularJS selector extentions
+`AngularJSSelector` contains a set of static methods to search by specified binding (`byModel`, `byBinding` and etc.).
```
```
`AngularJSSelector` contains a set of static methods to search for an HTML element by the specified binding (`byModel`, `byBinding` and etc.).
```

Commit:
fix Vasily's remarks

miherlosev commented:
ping @kirovboris

AlexanderMoskovkin commented on README.md:
```diff
@@ -10,88 +11,35 @@ npm install testcafe-angular-selectors
+fixture `App tests`
+    .page('http://angular-app-url')
+    .beforeEach(async t => {
+        await t.expect(waitForAngular()).ok();
```
If we name it as `waitForAngular` I expect to use it like:

```js
await waitForAngular(); // optionally waitForAngular(ms)
```

With the current implementation it should be something like `angularComponentsAreLoaded`. I'd prefer the first way, it's more easy to use. In this way we should add timeout logic into the ClientFunction.

What do you think @miherlosev , @kirovboris

miherlosev replied:
I agree with your proposal.
With `await waitForAngular()` is possible to raise a clear error message.

AlexanderMoskovkin commented on angular-selector.md:
````diff
@@ -0,0 +1,78 @@
+
+```
+
+See more examples [here](test/angular-selector-test.js);
````
It should be `.` at the end of line instead of `;`

Commit:
made links for Selector extention topics more notable

Commit:
increased timeout for 'waitForAngular' check in tests

miherlosev commented:
/cc @VasilyStrelyaev @kirovboris

AlexanderMoskovkin commented on test/angular-selector-test.js:
```diff
@@ -0,0 +1,47 @@
+test('composite selector', async t => {
+    const listItem           = AngularSelector('list list-item');
+    const listItemAngular6   = await listItem.nth(5).getAngular();
+    const listItemAngular5Id = listItem.nth(4).getAngular(({ state }) => state.id);
```
It seems `getAngular` method is not described in the docs

miherlosev replied:
Yes, you are right

kirovboris commented on README.md:
```diff
@@ -1,6 +1,7 @@
 
-This plugin provides selector extensions that make it easier to test Angular applications with [TestCafe](https://github.com/DevExpress/testcafe/). These extensions allow you to create a Selector to find elements on the page in a way that is native to Angular applications (like `binding`, `model` etc.. ).
+This plugin provides [Selector](https://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/selectors.html) extensions that make it easier to test Angular applications with [TestCafe](https://github.com/DevExpress/testcafe/). 
+These extensions allow you to create a Selector to find elements on the page in a way that is native to Angular applications.
```
`Selector`

kirovboris commented on angular-selector.md:
````diff
@@ -0,0 +1,78 @@
+const listItemComponent = AngularSelector('list list-item');
+```
+
+You can combine Angular selectors with TestCafe's Selector filter functions like .find, .withText, .nth and [other](http://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/selectors.html#functional-style-selectors).
````
Highlight Selector API methods(`.find`, `.withText` etc.) and add links to docs.

VasilyStrelyaev replied:
```
You can combine Angular selectors with TestCafe's Selector filter functions like .find, .withText, .nth and [others](http://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/selectors.html#functional-style-selectors).
```

miherlosev replied:
> and add links to docs.

I will not do this. We already use similar phrase in vue selectors
'You can combine Vue selectors with testcafe Selector filter functions like .find, .withText, .nth and other'.

kirovboris commented on src/angular-selector.js:
```diff
@@ -0,0 +1,70 @@
+
+    validateSelector(complexSelector);
+
+    // NOTE: We support only applications with a single root element
```
Why? We can look into every root component or make root component optional.

miherlosev replied:
I've investigated this area and didnt't find angular application with multiple roots.
So, now we don't support applications with multiple roots

AlexanderMoskovkin replied:
The phrase `We don't support an application with multiple roots` means that you can't use this plugin for such applications. In fact, you can. We have a note about this in the readme: `If your application has multiple roots, AngularSelector will return the first root returned by the window.getAllAngularRootElements function`.
So it's better to write comment like `If there are multiple roots on the page we find a target in the first root only`.

miherlosev replied:
Ok

kirovboris reviewed (changes requested):
/r-

VasilyStrelyaev commented on README.md:
```diff
@@ -10,88 +11,36 @@ npm install testcafe-angular-selectors
+This module includes separate helpers for AngularJS(1.x) and Angular(2+) applications.
 
-See more examples [here](/test/angular1-test.js).
+See the following topic for more details:
```
```
See the following topics for more details:
```

Commit:
fix review issues

Commit:
fix readme

miherlosev commented:
Changes:
- rewrote `waitForAngular` method, changed description
- described `AngularSelector.getAngular()` method

FPR @AlexanderMoskovkin @VasilyStrelyaev @kirovboris

VasilyStrelyaev commented on angular-selector.md:
```diff
@@ -0,0 +1,104 @@
+
+## Usage
+
+### Wait for application is ready to run tests
```
Wait for application to be ready to run tests

VasilyStrelyaev commented on angular-selector.md:
```diff
@@ -0,0 +1,104 @@
+
+As an alternative to [TestCafe snapshot properties](http://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/dom-node-state.html), you can obtain state of a Angular component.
+
+To obtain component state, use the Angular selector's .getAngular() method.
```
```
As an alternative to [TestCafe snapshot properties](http://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/dom-node-state.html), you can obtain the state of an Angular component.

To obtain the component state, use the Angular selector's `.getAngular()` method.
```

VasilyStrelyaev commented on angular-selector.md:
```diff
@@ -0,0 +1,104 @@
+
+### Wait for application is ready to run tests
+
+To wait until the Angular's component tree is loaded, add the `waitForAngular` method into fixture's `beforeEach` hook.
```
```
To wait until the Angular's component tree is loaded, add the `waitForAngular` method to fixture's `beforeEach` hook.
```

VasilyStrelyaev reviewed (changes requested):
Before @kirovboris said it: backtick all API members in the paragraphs you've added.

Commit:
small fixes

miherlosev commented:
@VasilyStrelyaev FPR

VasilyStrelyaev commented on angular-selector.md:
```diff
@@ -75,9 +75,9 @@ See more examples [here](test/angular-selector-test.js).
-To obtain component state, use the Angular selector's .getAngular() method.
+To obtain the component state, use the Angular selector's `.getAngular()` method.
 
 The .getAngular() method returns a [client function](https://devexpress.github.io/testcafe/documentation/test-api/obtaining-data-from-the-client.html). This function resolves to an object that contains component's state.
```
> .getAngular()

Some more things to backtick

VasilyStrelyaev commented on angular-selector.md:
```diff
@@ -0,0 +1,104 @@
+
+To obtain the component state, use the Angular selector's `.getAngular()` method.
+
+The .getAngular() method returns a [client function](https://devexpress.github.io/testcafe/documentation/test-api/obtaining-data-from-the-client.html). This function resolves to an object that contains component's state.
```
> .getAngular()

More things to backtick

VasilyStrelyaev commented on angular-selector.md:
````diff
@@ -0,0 +1,104 @@
+await t.expect(listAngular.testProp).eql(1);
+```
+
+As an alternative, the .getAngular() method can take a function that returns the required state property. 
````
And one more

VasilyStrelyaev commented on angular-selector.md:
````diff
@@ -0,0 +1,104 @@
+```
+
+As an alternative, the .getAngular() method can take a function that returns the required state property. 
+This function acts as a filter. Its argument is an object returned by .getAngular(), i.e. { state: ...}.
````
`.getAngular()`, i.e. `{ state: ...}`.

Two more

Commit:
small fixes

miherlosev commented:
FPR @kirovboris @AlexanderMoskovkin

kirovboris commented on test/wait-for-angular-test.js:
```diff
@@ -0,0 +1,43 @@
+fixture `Wait for Angular`;
+
+test('default timeout', async t => {
+    const start = +new Date();
```
typo in `+new` , here and further.

kirovboris commented on test/angular-selector-test.js:
```diff
@@ -0,0 +1,47 @@
+
+fixture `AngularSelector`
+    .page('http://localhost:8080/test/data/angular')
+    .beforeEach(async () => {
```
Why not just 
```js
.beforeEach(waitForAngular)
```

miherlosev replied:
It will not work.

kirovboris reviewed (changes requested):
/r-

AlexanderMoskovkin commented on angular-selector.md:
````diff
@@ -0,0 +1,104 @@
+```js
+import { AngularSelector } from 'testcafe-angular-selectors';
+
+const listConponent = AngularSelector('list');
````
a typo in the `listConponent`

AlexanderMoskovkin commented on angular-selector.md:
```diff
@@ -0,0 +1,104 @@
+
+### Obtaining component's state
+
+As an alternative to [TestCafe snapshot properties](http://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/dom-node-state.html), you can obtain the state of an Angular component.
```
We use `DOM Node State properties` instead of `snapshot` properties

AlexanderMoskovkin commented on src/angular-selector.js:
```diff
@@ -0,0 +1,70 @@
+        return rootElement;
+
+    const tags             = getTagList(complexSelector);
+    const debugElementRoot = window.ng.probe(rootElement);
```
Maybe we should raise an error with some meaningful description if `window.ng` is not defined? 
If an application loads slowly and a new user missed the `waitForAngular` function he sees the error like `Error in ClientFunction - window.ng is not defined`. And it's not clear what they should do.

How can we improve it:
If `window.ng` is not defined raise an error like `The page doesn't contain Angular components or they are not loaded completely or your Angular app in not in a development mode. Use the waitForAngular function to ensure the components are loaded.`
What do you think? /cc @VasilyStrelyaev @miherlosev

VasilyStrelyaev replied:
> in not in a development mode

typo

AlexanderMoskovkin commented on src/wait-for-angular.js:
```diff
@@ -0,0 +1,42 @@
+
+            return false;
+        };
+        const check                           = () => {
```
separate different multi-line var declarations with empty line, it'll improve readability.

AlexanderMoskovkin commented on test/data/angular/systemjs.config.js:
```diff
@@ -0,0 +1,28 @@
+            }
+        }
+    });
+})(this);
```
In the `testcafe` repo lint checks is there an empty line in the end of each file. I suppose we should do it here too

Commit:
fix Alexander's issues
